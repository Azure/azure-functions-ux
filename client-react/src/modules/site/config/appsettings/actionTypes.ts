export const AREA_STRING = '@appsettings';
export const APP_SETTINGS_FETCH_REQUEST = '@appsettings/FETCH_REQUEST';
export const APP_SETTINGS_FETCH_SUCCESS = '@appsettings/FETCH_SUCCESS';
export const APP_SETTINGS_FETCH_FAILURE = '@appsettings/FETCH_FAILURE';

export const APP_SETTINGS_UPDATE_REQUEST = '@appsettings/UPDATE_REQUEST';
export const APP_SETTINGS_UPDATE_SUCCESS = '@appsettings/UPDATE_SUCCESS';
export const APP_SETTINGS_UPDATE_FAILURE = '@appsettings/UPDATE_FAILURE';
