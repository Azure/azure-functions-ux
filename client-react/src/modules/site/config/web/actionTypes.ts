export const AREA_STRING = '@webconfig';
export const WEB_CONFIG_FETCH_REQUEST = '@webconfig/FETCH_REQUEST';
export const WEB_CONFIG_FETCH_SUCCESS = '@webconfig/FETCH_SUCCESS';
export const WEB_CONFIG_FETCH_FAILURE = '@webconfig/FETCH_FAILURE';

export const WEB_CONFIG_UPDATE_REQUEST = '@webconfig/UPDATE_REQUEST';
export const WEB_CONFIG_UPDATE_SUCCESS = '@webconfig/UPDATE_SUCCESS';
export const WEB_CONFIG_UPDATE_FAILURE = '@webconfig/UPDATE_FAILURE';
