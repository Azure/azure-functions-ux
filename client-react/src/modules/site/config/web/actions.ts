import { createStandardAction } from 'typesafe-actions';

import { ArmObj, SiteConfig } from '../../../../models/WebAppModels';
import {
  WEB_CONFIG_FETCH_FAILURE,
  WEB_CONFIG_FETCH_REQUEST,
  WEB_CONFIG_FETCH_SUCCESS,
  WEB_CONFIG_UPDATE_FAILURE,
  WEB_CONFIG_UPDATE_REQUEST,
  WEB_CONFIG_UPDATE_SUCCESS,
} from './actionTypes';

export const fetchWebConfigRequest = createStandardAction(WEB_CONFIG_FETCH_REQUEST)();
export const fetchWebConfigSuccess = createStandardAction(WEB_CONFIG_FETCH_SUCCESS).map((webConfig: ArmObj<SiteConfig>) => ({
  webConfig,
}));
export const fetchWebConfigFailure = createStandardAction(WEB_CONFIG_FETCH_FAILURE).map((error: Error) => ({
  error,
}));

export const updateWebConfigRequest = createStandardAction(WEB_CONFIG_UPDATE_REQUEST).map((webConfig: ArmObj<SiteConfig>) => ({
  webConfig,
}));
export const updateWebConfigSuccess = createStandardAction(WEB_CONFIG_UPDATE_SUCCESS).map((webConfig: ArmObj<SiteConfig>) => ({
  webConfig,
}));
export const updateWebConfigFailure = createStandardAction(WEB_CONFIG_UPDATE_FAILURE).map((error: Error) => ({
  error,
}));
