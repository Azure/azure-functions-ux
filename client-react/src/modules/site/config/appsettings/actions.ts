import { IAction } from '../../../../models/action';
import { IAppSettingsState } from './reducer';

export const UPDATE_SITE_APP_SETTINGS = 'UPDATE_SITE_APP_SETTINGS';
export const updateCurrentSiteAppSettings = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});

export const UPDATE_SITE_APP_SETTINGS_NO_CACHE = 'UPDATE_SITE_APP_SETTINGS_NO_CACHE';
export const updateCurrentSiteAppSettingsNoCache = (settings: Partial<IAppSettingsState>): IAction<Partial<IAppSettingsState>> => ({
  payload: settings,
  type: UPDATE_SITE_APP_SETTINGS,
});
