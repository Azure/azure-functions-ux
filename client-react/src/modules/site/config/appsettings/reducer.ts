import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import { ArmObj } from '../../../../models/WebAppModels';
import { metadataReducer } from '../../../ApiReducerHelper';
import { ApiState } from '../../../types';
import * as actions from './actions';
import { APP_SETTINGS_FETCH_SUCCESS, APP_SETTINGS_UPDATE_SUCCESS, AREA_STRING, UPDATE_APP_SETTINGS_FROM_SITE_UPDATE } from './actionTypes';

export type AppSettings = { [key: string]: string };

export type AppSettingsActions = ActionType<typeof actions>;
export type AppSettingsState = ApiState<ArmObj<AppSettings>>;

export const InitialState = {
  data: {
    id: '',
    properties: {},
    name: '',
    location: '',
    kind: '',
  },
};

export default combineReducers<AppSettingsState, AppSettingsActions>({
  metadata: metadataReducer(AREA_STRING),
  data: (state = InitialState.data, action) => {
    switch (action.type) {
      case APP_SETTINGS_FETCH_SUCCESS:
        return action.appSettings;
      case APP_SETTINGS_UPDATE_SUCCESS:
        return action.appSettings;
      case UPDATE_APP_SETTINGS_FROM_SITE_UPDATE:
        return { ...state, properties: action.appSettings ? action.appSettings : state.properties };
      default:
        return state;
    }
  },
});
