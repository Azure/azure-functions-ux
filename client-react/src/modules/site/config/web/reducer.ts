import { ArmObj, SiteConfig, VirtualApplication } from '../../../../models/WebAppModels';
import { IAction } from '../../../../models/action';
import { UPDATE_SITE_CONFIG_WEB, UPDATE_SITE_CONFIG_WEB_NO_CACHE } from './actions';
import { DEFAULT_KEY, generateCacheTTL } from 'redux-cache';

export interface IWebConfigState {
  loading: boolean;
  config: ArmObj<SiteConfig>;
  virtualApplications: VirtualApplication[];
  currentlySelectedStack: string;
  saving: boolean;
}
export const InitialState: IWebConfigState = {
  [DEFAULT_KEY]: null,
  loading: false,
  saving: false,
  currentlySelectedStack: '',
  config: {
    id: '',
    properties: {} as any,
    name: '',
    location: '',
    kind: '',
  },
  virtualApplications: [],
};

const webConfig = (state = InitialState, action: IAction<Partial<IWebConfigState>>) => {
  switch (action.type) {
    case UPDATE_SITE_CONFIG_WEB:
      return {
        ...state,
        [DEFAULT_KEY]: generateCacheTTL(6000),
        ...action.payload,
      };
    case UPDATE_SITE_CONFIG_WEB_NO_CACHE:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

export default webConfig;
