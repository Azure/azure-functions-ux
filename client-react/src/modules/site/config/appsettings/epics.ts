import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import * as Types from '../../../types';
import {
  fetchAppSettingsFailure,
  fetchAppSettingsRequest,
  fetchAppSettingsSuccess,
  updateAppSettingsFailure,
  updateAppSettingsRequest,
  updateAppSettingsSuccess,
} from './actions';
import { AppSettingsActions } from './reducer';

export const fetchAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(fetchAppSettingsRequest)),
    switchMap(() =>
      from(appSettingsApi.fetchAppSettings(store.value)).pipe(
        map(fetchAppSettingsSuccess),
        catchError(err => of(fetchAppSettingsFailure(err)))
      )
    )
  );

export const updateAppSettings: Epic<AppSettingsActions, AppSettingsActions, Types.RootState, Types.Services> = (
  action$,
  store,
  { appSettingsApi }
) =>
  action$.pipe(
    filter(isActionOf(updateAppSettingsRequest)),
    switchMap(action =>
      from(appSettingsApi.updateAppSettings(store.value, action.appSettings)).pipe(
        map(updateAppSettingsSuccess),
        catchError(err => of(updateAppSettingsFailure(err)))
      )
    )
  );
export default combineEpics(fetchAppSettings, updateAppSettings);
