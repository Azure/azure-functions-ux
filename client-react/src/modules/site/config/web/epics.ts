import { combineEpics, Epic } from 'redux-observable';
import { from, of } from 'rxjs';
import { catchError, filter, map, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import * as Types from '../../../types';
import {
  fetchWebConfigFailure,
  fetchWebConfigRequest,
  fetchWebConfigSuccess,
  updateWebConfigFailure,
  updateWebConfigRequest,
  updateWebConfigSuccess,
} from './actions';
import { ConfigAction } from './reducer';

export const fetchWebConfig: Epic<ConfigAction, ConfigAction, Types.RootState, Types.Services> = (action$, store, { webConfigApi }) =>
  action$.pipe(
    filter(isActionOf(fetchWebConfigRequest)),
    switchMap(() =>
      from(webConfigApi.fetchWebConfig(store.value)).pipe(
        map(fetchWebConfigSuccess),
        catchError(err => of(fetchWebConfigFailure(err)))
      )
    )
  );

export const updateWebConfig: Epic<ConfigAction, ConfigAction, Types.RootState, Types.Services> = (action$, store, { webConfigApi }) =>
  action$.pipe(
    filter(isActionOf(updateWebConfigRequest)),
    switchMap(action =>
      from(webConfigApi.updateWebConfig(store.value, action.webConfig)).pipe(
        map(updateWebConfigSuccess),
        catchError(err => of(updateWebConfigFailure(err)))
      )
    )
  );
export default combineEpics(fetchWebConfig, updateWebConfig);
