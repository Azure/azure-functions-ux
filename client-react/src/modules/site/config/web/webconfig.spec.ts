import { ActionsObservable } from 'redux-observable';
import { toArray } from 'rxjs/operators';

import rootReducer from '../../..';
import { IStartupInfo } from '../../../../models/portal-models';
import { ArmObj, SiteConfig } from '../../../../models/WebAppModels';
import { getStartupInfoAction } from '../../../portal/actions';
import { RootState, Services } from '../../../types';
import { updateResourceId } from '../../actions';
import {
  fetchWebConfigFailure,
  fetchWebConfigRequest,
  fetchWebConfigSuccess,
  updateWebConfigFailure,
  updateWebConfigRequest,
  updateWebConfigSuccess,
} from './actions';
import { WEB_CONFIG_FETCH_FAILURE, WEB_CONFIG_FETCH_SUCCESS, WEB_CONFIG_UPDATE_FAILURE, WEB_CONFIG_UPDATE_SUCCESS } from './actionTypes';
import { fetchWebConfig, updateWebConfig } from './epics';
import reducer from './reducer';
import api from './webConfigApiService';
jest.mock('../../../ArmHelper');
import MakeArmCall from '../../../ArmHelper';

const testResult: ArmObj<SiteConfig> = {
  id: '',
  name: '',
  location: '',
  kind: '',
  properties: {
    scmType: 'testvalue',
  },
} as ArmObj<SiteConfig>;
describe('Web Config Names Store Epics', () => {
  const successDeps = {
    webConfigApi: {
      fetchWebConfig: async (state: RootState): Promise<ArmObj<SiteConfig>> => {
        return testResult;
      },
      updateWebConfig: async (state: RootState, newConfig: ArmObj<SiteConfig>): Promise<ArmObj<SiteConfig>> => {
        return testResult;
      },
    },
  } as Services;

  const failDeps = {
    webConfigApi: {
      fetchWebConfig: async (state: RootState): Promise<ArmObj<SiteConfig>> => {
        throw new Error('failuremessage');
      },
      updateWebConfig: async (state: RootState, newConfig: ArmObj<SiteConfig>): Promise<ArmObj<SiteConfig>> => {
        throw new Error('failuremessage');
      },
    },
  } as Services;

  it('Sends Success Action with object on Successful Fetch', async () => {
    let action$ = ActionsObservable.of(fetchWebConfigRequest());
    const output$ = fetchWebConfig(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(WEB_CONFIG_FETCH_SUCCESS);
    if (action.type === WEB_CONFIG_FETCH_SUCCESS) {
      expect(action.webConfig.properties.scmType).toBe('testvalue');
    }
  });

  it('Sends Success Action with object on Successful Update', async () => {
    let action$ = ActionsObservable.of(updateWebConfigRequest(testResult));
    const output$ = updateWebConfig(action$, {} as any, successDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(WEB_CONFIG_UPDATE_SUCCESS);
    if (action.type === WEB_CONFIG_UPDATE_SUCCESS) {
      expect(action.webConfig.properties.scmType).toBe('testvalue');
    }
  });

  it('Sends Error Action with error on failed Fetch', async () => {
    let action$ = ActionsObservable.of(fetchWebConfigRequest());
    const output$ = fetchWebConfig(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(WEB_CONFIG_FETCH_FAILURE);
    if (action.type === WEB_CONFIG_FETCH_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });

  it('Sends Error Action with error on failed Update', async () => {
    let action$ = ActionsObservable.of(updateWebConfigRequest(testResult));
    const output$ = updateWebConfig(action$, {} as any, failDeps);
    const actions = await output$.pipe(toArray()).toPromise();
    expect(actions.length).toBe(1);
    const action = actions[0];
    expect(action.type).toBe(WEB_CONFIG_UPDATE_FAILURE);
    if (action.type === WEB_CONFIG_UPDATE_FAILURE) {
      expect(action.error.message).toBe('failuremessage');
    }
  });
});

describe('Web Config Names Store Reducer', () => {
  const initialState = reducer(undefined, {} as any);
  describe('initial state', () => {
    it('should match a snapshot', () => {
      expect(initialState).toMatchSnapshot();
    });

    it('loading and updating should be false', () => {
      expect(initialState.metadata.loading).toBe(false);
      expect(initialState.metadata.updating).toBe(false);
    });
  });

  describe('Web Config Names Fetch Stories', () => {
    it('should trigger loading when the fetch is requested', () => {
      const action = fetchWebConfigRequest();
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(true);
    });

    it('metadata and app settings should update on successful load', () => {
      const action = fetchWebConfigSuccess(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.data.properties.scmType).toBe('testvalue');
    });

    it('error should be reflected on failed load', () => {
      const action = fetchWebConfigFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.loading).toBe(false);
      expect(state.metadata.fetchError).toBe(true);
      expect(state.metadata.fetchErrorObject.message).toBe('testerror');
    });
  });

  describe('Web Config Names Update Stories', () => {
    it('should trigger updating when the update is requested', () => {
      const action = updateWebConfigRequest(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(true);
    });

    it('metadata and app settings should update on successful load', () => {
      const action = updateWebConfigSuccess(testResult);
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.data.properties.scmType).toBe('testvalue');
    });

    it('error should be reflected on failed load', () => {
      const action = updateWebConfigFailure(new Error('testerror'));
      const state = reducer(initialState, action);
      expect(state.metadata.updating).toBe(false);
      expect(state.metadata.updateError).toBe(true);
      expect(state.metadata.updateErrorObject.message).toBe('testerror');
    });
  });
});

describe('Web Config Names Service', () => {
  const initialState = rootReducer(undefined, {} as any);
  let state;
  beforeEach(() => {
    const updateResourceIdAction = updateResourceId('resourceid');
    const updateSUIAction = getStartupInfoAction({
      token: 'testtoken',
      armEndpoint: 'testEndpoint',
    } as IStartupInfo);

    state = rootReducer(rootReducer(initialState, updateResourceIdAction), updateSUIAction);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Fetch Api calls api with appropriate info', async () => {
    api.fetchWebConfig(state);
    expect(MakeArmCall).toHaveBeenCalledWith('testEndpoint', 'testtoken', 'resourceid/config/web', 'FetchWebConfig');
  });

  it('Update Api calls api with appropriate info', async () => {
    api.updateWebConfig(state, testResult);
    expect(MakeArmCall).toHaveBeenCalledWith('testEndpoint', 'testtoken', 'resourceid/config/web', 'UpdateWebConfig', 'PUT', testResult);
  });
});
