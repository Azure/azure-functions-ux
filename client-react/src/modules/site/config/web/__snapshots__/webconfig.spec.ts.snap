// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Config Names Store Reducer initial state should match a snapshot 1`] = `
Object {
  "data": Object {
    "id": "",
    "kind": "",
    "location": "",
    "name": "",
    "properties": Object {},
  },
  "metadata": Object {
    "fetchError": false,
    "fetchErrorObject": [Error],
    "loading": false,
    "updateError": false,
    "updateErrorObject": [Error],
    "updating": false,
  },
}
`;
